### PR TITLE
Only run workflows if they might be able to work

### LINKED ISSUE #63163: CI fails in forks because it uses secrets unconditionally
**What happened**:

Lots of workflows fail when triggered in forks

**What you expected to happen**:

* workflows should declare permissions they need
* workflows should verify they have any secrets they need before trying to do work that requires them

**How to reproduce it (as minimally and precisely as possible)**:

1. create an organization
2. set the organization permissions to readonly instead of read write https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#configuring-the-default-github_token-permissions
3. create a fork of the grafana repository
4. go to actions and enable actions (there are ways that they'll be automatically triggered)
5. you can just wait for a scheduled task to run, or you can trigger various events (e.g. creating an issue in your fork w/ a label, or creating a PR in your fork)

**Anything else we need to know?**:

**Environment**:
- Grafana version: main
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by jsoref**
**What is this feature?**

Check for secrets in workflows that use secrets

**Why do we need this feature?**

Otherwise forks get lots of failures

**Who is this feature for?**

Contributors who aren't members of the grafana organization and thus need to use forks where github actions may run, and especially those who run those forks in organizations that are moderately concerned about security (and thus have restricted the github_token's default permissions)

**Which issue(s) does this PR fix?**:

Fixes #63163

**Special notes for your reviewer**:

This is a followup to #63164 as some workflows changed since it was written.
@armandgrillet

### CODE REVIEW

**Comment by jsoref** [comment]
@joshhunt

File: .github/workflows/detect-breaking-changes-report.yml [hunk 1]
```diff
@@ -15,6 +15,11 @@ jobs:
     env:
       ARTIFACT_NAME: 'levitate' # The name of the artifact that we would like to download
       ARTIFACT_FOLDER: '${{ github.workspace }}/tmp' # The name of the folder where we will download the artifact to
+    permissions:
+      contents: write
+      issues: write
+      pull-requests: write
+      artifacts: read
```
Comment (jsoref) [inline .github/workflows/detect-breaking-changes-report.yml:22]: this line caused issues. I need to figure out the story

File: .github/workflows/commands.yml [hunk 2]
```diff
@@ -7,7 +7,22 @@ on:
...
+  config:
+    runs-on: "ubuntu-latest"
+    outputs:
+      has-secrets: ${{ steps.check.outputs.has-secrets }}
+    steps:
+      - name: "Check for secrets"
+        id: check
+        shell: bash
+        run: |
+          if [ -n "${{ (secrets.GRAFANA_MISC_STATS_API_KEY != '' && secrets.GH_BOT_ACCESS_TOKEN != '') || '' }}" ]; then
```
Comment (joshhunt) [inline .github/workflows/commands.yml:19]:
```suggestion
          if [ -n "${{ (secrets.GRAFANA_MISC_STATS_API_KEY != '' && secrets.ISSUE_COMMANDS_TOKEN != '') || '' }}" ]; then
```

I think this should reference the `ISSUE_COMMANDS_TOKEN` secret instead?

File: .github/workflows/community-release.yml [hunk 3]
```diff
@@ -6,7 +6,26 @@ on:
         required: true
         description: 'Needs to match, exactly, the name of a milestone. The version to be released please respect: major.minor.patch, major.minor.patch-preview or major.minor.patch-preview<number> format. example: 7.4.3, 7.4.3-preview or 7.4.3-preview1'
 jobs:
+  config:
```
Comment (joshhunt) [inline .github/workflows/community-release.yml:9]: I think we can revert this - this change only runs manually via `workflow_dispatch`, and the less actions yaml to maintain the better

File: .github/workflows/detect-breaking-changes-report.yml [hunk 4]
```diff
@@ -132,13 +136,15 @@ jobs:
...
+        HAS_SECRETS: ${{ (github.repository == "grafana/grafana" || secrets.SLACK_LEVITATE_WEBHOOK_URL != '') || '' }}
 
     # Add the label
     - name: Add "levitate breaking change" label
-      if: steps.levitate-run.outputs.exit_code == 1 && steps.does-label-exist.outputs.result == 0 && steps.levitate-run.outputs.shouldSkip != 'true'
+      if: steps.levitate-run.outputs.exit_code == 1 && steps.does-label-exist.outputs.result == 0 && steps.levitate-run.outputs.shouldSkip != 'true' && env.HAS_SECRETS
       uses: actions/github-script@v6
       env:
         PR_NUMBER: ${{ steps.levitate-run.outputs.pr_number }}
+        HAS_SECRETS: ${{ (secrets.GH_BOT_ACCESS_TOKEN != '') || '' }}
```
Comment (joshhunt) [inline .github/workflows/detect-breaking-changes-report.yml:147]: `detect-breaking-changes-report.yml` now uses `GITHUB_TOKEN`, which will always be defined, so I think this environment variable can be removed?

File: .github/workflows/detect-breaking-changes-report.yml [hunk 5]
```diff
@@ -132,13 +136,15 @@ jobs:
           }
       env:
         SLACK_WEBHOOK_URL: ${{ secrets.SLACK_LEVITATE_WEBHOOK_URL }}
+        HAS_SECRETS: ${{ (github.repository == "grafana/grafana" || secrets.SLACK_LEVITATE_WEBHOOK_URL != '') || '' }}
```
Comment (joshhunt) [inline .github/workflows/detect-breaking-changes-report.yml:139]: this one should stay i guess

File: .github/workflows/detect-breaking-changes-report.yml [hunk 6]
```diff
@@ -151,10 +157,11 @@ jobs:
 
     # Remove label (no more breaking changes)
     - name: Remove "levitate breaking change" label
-      if: steps.levitate-run.outputs.exit_code == 0 && steps.does-label-exist.outputs.result == 1 && steps.levitate-run.outputs.shouldSkip != 'true'
+      if: steps.levitate-run.outputs.exit_code == 0 && steps.does-label-exist.outputs.result == 1 && steps.levitate-run.outputs.shouldSkip != 'true' && env.HAS_SECRETS
       uses: actions/github-script@v6
       env:
         PR_NUMBER: ${{ steps.levitate-run.outputs.pr_number }}
+        HAS_SECRETS: ${{ (secrets.GH_BOT_ACCESS_TOKEN != '') || '' }}
```
Comment (joshhunt) [inline .github/workflows/detect-breaking-changes-report.yml:164]: `detect-breaking-changes-report.yml` now uses `GITHUB_TOKEN`, which will always be defined, so I think this environment variable can be removed?

File: .github/workflows/detect-breaking-changes-report.yml [hunk 7]
```diff
@@ -169,10 +176,11 @@ jobs:
     # This is very weird, the actual request goes through (comes back with a 201), but does not assign the team.
     # Related issue: https://github.com/renovatebot/renovate/issues/1908
     - name: Add "grafana/plugins-platform-frontend" as a reviewer
-      if: steps.levitate-run.outputs.exit_code && steps.levitate-run.outputs.shouldSkip != 'true'
+      if: steps.levitate-run.outputs.exit_code && steps.levitate-run.outputs.shouldSkip != 'true' && env.HAS_SECRETS
       uses: actions/github-script@v6
       env:
         PR_NUMBER: ${{ steps.levitate-run.outputs.pr_number }}
+        HAS_SECRETS: ${{ (secrets.GH_BOT_ACCESS_TOKEN != '') || '' }}
```
Comment (joshhunt) [inline .github/workflows/detect-breaking-changes-report.yml:183]: `detect-breaking-changes-report.yml` now uses `GITHUB_TOKEN`, which will always be defined, so I think this environment variable can be removed?

File: .github/workflows/detect-breaking-changes-report.yml [hunk 8]
```diff
@@ -186,10 +194,11 @@ jobs:
 
     # Remove reviewers (no more breaking changes)
     - name: Remove "grafana/plugins-platform-frontend" from the list of reviewers
-      if: steps.levitate-run.outputs.exit_code == 0 && steps.levitate-run.outputs.shouldSkip != 'true'
+      if: steps.levitate-run.outputs.exit_code == 0 && steps.levitate-run.outputs.shouldSkip != 'true' && env.HAS_SECRETS
       uses: actions/github-script@v6
       env:
         PR_NUMBER: ${{ steps.levitate-run.outputs.pr_number }}
+        HAS_SECRETS: ${{ (secrets.GH_BOT_ACCESS_TOKEN != '') || '' }}
```
Comment (joshhunt) [inline .github/workflows/detect-breaking-changes-report.yml:201]: `detect-breaking-changes-report.yml` now uses `GITHUB_TOKEN`, which will always be defined, so I think this environment variable can be removed?

File: .github/workflows/detect-breaking-changes-report.yml [hunk 9]
```diff
@@ -15,6 +15,10 @@ jobs:
     env:
       ARTIFACT_NAME: 'levitate' # The name of the artifact that we would like to download
       ARTIFACT_FOLDER: '${{ github.workspace }}/tmp' # The name of the folder where we will download the artifact to
+    permissions:
+      contents: write
```
Comment (joshhunt) [inline .github/workflows/detect-breaking-changes-report.yml:19]: I don't think this action requires `contents: write`? I think it can get away with `contents: read` @grafana/plugins-platform-backend

File: .github/workflows/metrics-collector.yml [hunk 10]
```diff
@@ -16,7 +16,22 @@ on:
...
+  config:
+    runs-on: "ubuntu-latest"
+    outputs:
+      has-secrets: ${{ steps.check.outputs.has-secrets }}
+    steps:
+      - name: "Check for secrets"
+        id: check
+        shell: bash
+        run: |
+          if [ -n "${{ (secrets.GRAFANA_MISC_STATS_API_KEY != '' && secrets.GH_BOT_ACCESS_TOKEN != '') || '' }}" ]; then
```
Comment (joshhunt) [inline .github/workflows/metrics-collector.yml:28]: 👍🏻  this is fine - this is the one workflow remaining that still references `GH_BOT_ACCESS_TOKEN`

File: .github/workflows/milestone.yml [hunk 11]
```diff
@@ -6,12 +6,28 @@ on:
         description: 'The version to be released please respect: major.minor.patch, major.minor.patch-preview or major.minor.patch-preview<number> format. example: 7.4.3, 7.4.3-preview or 7.4.3-preview1'
         required: true
 jobs:
+  config:
```
Comment (joshhunt) [inline .github/workflows/milestone.yml:9]: Can we revert the changes to this file? This only runs manually.

File: .github/workflows/pr-checks.yml [hunk 12]
```diff
@@ -15,8 +15,13 @@ on:
       - demilestoned
 concurrency:
   group: pr-checks-${{ github.event.number }}
+permissions:
+  contents: write
 jobs:
   main:
+    permissions:
+      actions: write
+      contents: read
```
Comment (joshhunt) [inline .github/workflows/pr-checks.yml:24]: I think this check only requires:

```yaml
permissions:
  contents: read
  pull-requests: write
```

@marefr or @zerok might know more?

File: .github/workflows/pr-commands-closed.yml [hunk 13]
```diff
@@ -15,4 +17,4 @@ jobs:
         echo PR #${{ github.event.number }} has been closed without being merged, removing milestone.
         gh pr edit ${{ github.event.number }} --milestone "" --repo $GITHUB_REPOSITORY
       env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
\ No newline at end of file
+        GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (joshhunt) [inline .github/workflows/pr-commands-closed.yml:20]: I don't think changes is needed - `GITHUB_TOKEN` I think works just fine https://cli.github.com/manual/gh_help_environment

File: .github/workflows/pr-commands.yml [hunk 14]
```diff
@@ -8,7 +8,25 @@ on:
...
+    runs-on: "ubuntu-latest"
+    outputs:
+      has-secrets: ${{ steps.check.outputs.has-secrets }}
+    steps:
+      - name: "Check for secrets"
+        id: check
+        shell: bash
+        run: |
+          if [ -n "${{ (secrets.GRAFANA_PR_AUTOMATION_APP_ID != '' &&
+                        secrets.GRAFANA_DELIVERY_BOT_APP_PEM != '' &&
```
Comment (joshhunt) [inline .github/workflows/pr-commands.yml:21]:
```suggestion
                        secrets.GRAFANA_PR_AUTOMATION_APP_PEM != '' &&
```

File: .github/workflows/stale.yml [hunk 15]
```diff
@@ -7,7 +7,22 @@ permissions:
   pull-requests: write
 
 jobs:
+  config:
```
Comment (joshhunt) [inline .github/workflows/stale.yml:10]: action now uses `GITHUB_TOKEN`, so i think these changes should be reverted

**Review by joshhunt — COMMENTED** [review]
My pass 👍🏻

File: .github/workflows/pr-commands-closed.yml (hunk 13 above)
Comment (jsoref) [reply .github/workflows/pr-commands-closed.yml:20]: `GITHUB_TOKEN` isn't set by default https://docs.github.com/en/actions/learn-github-actions/variables (I've tripped on this)

File: .github/workflows/pr-checks.yml (hunk 12 above)
Comment (jsoref) [reply .github/workflows/pr-checks.yml:24]: My guess is this was the code that mattered:
https://github.com/grafana/grafana-github-actions/blob/6834fcbfb10a065f2fd6e20f04417252a9ca2397/pr-checks/Dispatcher.ts#L78-L83
https://docs.github.com/en/rest/commits/statuses?apiVersion=2022-11-28#create-a-status

File: .github/workflows/pr-commands-closed.yml (hunk 13 above)
Comment (jsoref) [reply .github/workflows/pr-commands-closed.yml:20]: Oh, sorry. Yes, both `GITHUB_TOKEN` and `GH_TOKEN` work, however there is a preference for `GH_TOKEN`.

I'm happy to drop this.

File: .github/workflows/pr-checks.yml (hunk 12 above)
Comment (jsoref) [reply .github/workflows/pr-checks.yml]: Dropped in https://github.com/grafana/grafana/pull/72503#event-9948642305: https://github.com/grafana/grafana/compare/89f76cea1c02015c1e6cbd9164e447584df2c2cd..5befafda553b4b7c4e46ed2af55fc772254f1dc9 (GitHub doesn't seem to be properly updating the PR state even though the push happened -- I trip on this sort of thing occasionally -- GitHub aims for "eventual consistency"...)

Comment (joshhunt) [reply .github/workflows/pr-checks.yml]: 👍🏻  It looks like `statuses: write` will be the correct permission for that then? 

https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs#overview

For clarity for others, by default it defaults to read/write https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

Comment (jsoref) [reply .github/workflows/pr-checks.yml]: If it's status that was the write, yes. I'm not really sure what I saw at the time.

Fwiw, the default permissions can be different by org/enterprise (my org where I do work defaults to `read` in general which is why this PR / issue exists): 
https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-organization-settings/disabling-or-limiting-github-actions-for-your-organization#configuring-the-default-github_token-permissions

Comment (joshhunt) [reply .github/workflows/pr-checks.yml]: Yup - I meant that more as a "we're not escalating _our_ default permissions, just making the defaults explicit" 👍🏻

**Review by jdbaldry — APPROVED** [review]
`publish-technical-documentation-*.yml` workflows LGTM! Thank you for your work here.

File: .github/workflows/detect-breaking-changes-report.yml (hunk 5 above)
Comment (joshhunt) [reply .github/workflows/detect-breaking-changes-report.yml:139]: FYI/TIL - Apparently strings must be in github actions expressions, strings must be quoted with _single_  quotes.

https://docs.github.com/en/actions/learn-github-actions/expressions#literals 

image

 I will fix this and push up.

**Review by joshhunt — APPROVED** [review]
👍 Looks great, thanks for the perseverance on this!

I've fixed the one issue with quotes, rebased, and force pushed up. We'll aim to merge this today 👏🏻

File: .github/workflows/pr-checks.yml (hunk 12 above)
Comment (jsoref) [reply .github/workflows/pr-checks.yml]: For the record, it appears that `statuses` was indeed necessary, on the bright side, we did talk about it. https://github.com/grafana/grafana/commit/3172715a025639e09b4474a0a3a855cadcd23f64
